[ENT-860]Update Basket and Receipt Page UI to include Bulk Enrollment Information

asadiqbal08 opened the pull request:
##### Updated the screen shots
[ENT-860](https://openedx.atlassian.net/browse/ENT-860)

**Testing Intructions**

- Visit [business sandbox](https://ecommerce-business.sandbox.edx.org/courses/) and create a new verified course in ecommerce with bulk enrollment enabled. 
- Enroll a user through LMS and visit basket page. 
- View the UI changes there.

##### NOTE: Currently, Getting some error while adding a course in ecommerce on sandbox. A ticket is already logged [here](https://openedx.atlassian.net/browse/ENT-890). In that case, Try to [Hit the URL with a new user](https://ecommerce-business.sandbox.edx.org/basket/single-item/?sku=019379F) to view the page directly that I configured manually for a course on sandbox.

**1- Notifications when update the quantity and jump back to single purchase after clicking update button. 
2- Added the organization field, with required field validation, when there is an enrollment code product in basket.** 

<img width="617" alt="1" src="https://user-images.githubusercontent.com/7334669/36587178-2b5533ee-18a6-11e8-9248-6e364d0b9c6d.png">

**3- Update quantity message in notification bar.** 

<img width="632" alt="quantity" src="https://user-images.githubusercontent.com/7334669/36587210-4980ab5a-18a6-11e8-86df-4151eb16b38c.png">

**4- Single purchase message in notification bar along with a link to course about page "Click here to enroll directly"**

<img width="603" alt="screen shot 2018-02-23 at 2 30 43 pm" src="https://user-images.githubusercontent.com/7334669/36587216-560ad7ec-18a6-11e8-81ff-6e01a88394c4.png">


**5- Receipt page changes.**

<img width="810" alt="receipt page" src="https://user-images.githubusercontent.com/7334669/36535491-ac09cea2-17eb-11e8-8cb8-12f5bea9629c.png">

**6- Organization is Required Field**

<img width="641" alt="screen shot 2018-02-22 at 3 50 48 pm" src="https://user-images.githubusercontent.com/7334669/36534393-3d31a318-17e8-11e8-96f0-7bb11053376d.png">

#### NOTE: We need to update the receipt page message as I posted a comment on [JIRA](https://openedx.atlassian.net/browse/ENT-860)

zubair-arbi commented on ecommerce/templates/oscar/basket/messages/new_total.html:
```diff
@@ -0,0 +1,20 @@
+{% load i18n %}
+{% load currency_filters %}
```
nit: remove this templatetag as we are not using it.

asadiqbal08 replied:
we are using it now.

zubair-arbi commented on ecommerce/templates/oscar/basket/messages/new_total.html:
```diff
@@ -0,0 +1,20 @@
+        {% trans "Your basket is now empty" %}
+    {% else %}
+        {% blocktrans with num_items=basket.num_items %}
+            <b>We’ve updated your quantity.</b>
```
move html tags outside the block `blocktrans` or use some substitute for it.
```html
{% blocktrans with link_start='<a href="https://www.edx.org/gfa">' link_end='</a>' %}
    Please read {{ link_start }}our FAQs to view common questions about our certificates.{{ link_end }}
{% endblocktrans %}
```

zubair-arbi commented on ecommerce/extensions/payment/forms.py:
```diff
@@ -100,6 +104,17 @@ def __init__(self, user, request, *args, **kwargs):
+                if basket_has_enrollment_code_product and 'organization' not in self.fields:
+                    # If basket has any enrollment code items then we will add an organization
+                    # field next to "last_name."
+                    self.fields['organization'] = forms.CharField(max_length=60, label=_('Organization'))
```
In actual `BusinessClient` model the name field has limit of 255 characters:
https://github.com/edx/ecommerce/blob/master/ecommerce/core/models.py#L591-L594
```python
class BusinessClient(models.Model):
    """The model for the business client."""
    name = models.CharField(_('Name'), unique=True, max_length=255)
```

zubair-arbi replied:
Also please add validation (empty organization field in case of enrollment code) of organization (business client) field in the form's `clean` method [here](https://github.com/edx/ecommerce/blob/master/ecommerce/extensions/payment/forms.py#L131-L182).

asadiqbal08 replied:
I did not see the validation on first_name, last_name in clean method. currently, There are two levels of validation. 
1- Client side 
2- backend
on backend side, The following code marking the `required` attribute fields as [HERE](https://github.com/edx/ecommerce/blob/master/ecommerce/extensions/payment/forms.py#L101). The payment form method form_invalid will raise the error if client side validation failed to handle.

asadiqbal08 commented:
@zubair-arbi kindly take another look.

asadiqbal08 commented:
@rrakhshan fyi
<img width="816" alt="screen shot 2018-02-23 at 11 11 24 am" src="https://user-images.githubusercontent.com/7334669/36580512-629fe284-188a-11e8-9bac-cc90c656aace.png">

brittneyexline commented:
@asadiqbal08 can you provide testing instructions for the business sandbox?

asadiqbal08 commented:
@brittneyexline added.

brittneyexline commented on ecommerce/core/url_utils.py:
```diff
@@ -47,6 +47,23 @@ def get_lms_courseware_url(course_run_id):
+    return get_lms_url('courses/{}/about'.format(course_key))
+
+
+def get_lms_explore_courses():
```
to be consistent, rename to get_lms_explore_courses_url

brittneyexline commented on ecommerce/extensions/payment/tests/test_forms.py:
```diff
@@ -132,3 +162,42 @@ def test_countries_sorting(self):
         self.assertEqual(actual, expected)
+
+    def test_organization_field_in_form(self):
+        """ Verify the field 'organization' is present in form when basket have enrollment code product. """
```
"Verify the field 'organization' is present in the form when the basket has an enrollment code product."

brittneyexline commented on ecommerce/extensions/payment/tests/test_forms.py:
```diff
@@ -132,3 +162,42 @@ def test_countries_sorting(self):
+        self.assertTrue('organization' in form.fields)
+
+    def test_organization_field_not_in_form(self):
+        """ Verify the field 'organization' is not present in form when basket has not any enrollment code product. """
```
"Verify the field 'organization' is not present in the form when the basket does not have an enrollment code product."

brittneyexline commented on ecommerce/templates/edx/checkout/receipt.html:
```diff
@@ -165,6 +172,11 @@ <h2>{% trans "Order Information" %}</h2>
         </div>
       {% endif %}
+      <div id="find-more-courses-link">
+          <a class="find-more-courses-link nav-link" href="{{ explore_courses_url }}" target="_blank">
```
is this meant to open in a new tab?

asadiqbal08 replied:
yes

asadiqbal08 commented:
@brittneyexline I have addressed some of your suggestions. I saw your comment about "Verify your Identity" box. The changes in the PR are not related to that box. Can we open a new ticket for any changes related to that particular box?
